File: singlylist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insert_at_beginning(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def insert_at_end(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node

    def insert_in_mid(self, data, key):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        temp = self.head
        while temp and temp.data != key:
            temp = temp.next
        if temp is None:
            print("Key not found in the list")
            return
        new_node.next = temp.next
        temp.next = new_node

    def delete_node(self, key):
        temp = self.head
        if temp is not None and temp.data == key:
            self.head = temp.next
            temp = None
            return
        prev = None
        while temp and temp.data != key:
            prev = temp
            temp = temp.next
        if temp is None:
            print("Key not found in the list")
            return
        prev.next = temp.next
        temp = None

    def print_list(self):
        temp = self.head
        while temp:
            print(temp.data, end="-")
            temp = temp.next
        print()

list = LinkedList()

# Insert nodes
list.insert_at_end(1)
list.insert_at_end(2)
list.insert_at_end(3)
list.insert_at_end(4)

print("Initial List: ", end="")
list.print_list()

print("After inserting in mid: ", end="")
list.insert_in_mid(5, 6)
list.print_list()

# Delete a node
print("After deleting 3: ", end="")
list.delete_node(2)
list.print_list()

# Insert at beginning
print("After inserting 5 at beginning: ", end="")
list.insert_at_beginning(0)
list.print_list()

File: singlylist.cpp
#include <iostream>

using namespace std;

class Node {
public:
    int data;
    Node* next;

    Node(int data) {
        this->data = data;
        this->next = nullptr;
    }
};

class LinkedList {
public:
    Node* head;

    LinkedList() {
        this->head = nullptr;
    }

    void insertAtBeginning(int data) {
        Node* new_node = new Node(data);
        new_node->next = head;
        head = new_node;
    }

    void insertAtEnd(int data) {
        Node* new_node = new Node(data);
        if (head == nullptr) {
            head = new_node;
            return;
        }
        Node* temp = head;
        while (temp->next != nullptr) {
            temp = temp->next;
        }
        temp->next = new_node;
    }

    void insertInMid(int data, int key) {
        Node* new_node = new Node(data);
        if (head == nullptr) {
            head = new_node;
            return;
        }
        Node* temp = head;
        while (temp != nullptr && temp->data != key) {
            temp = temp->next;
        }
        if (temp == nullptr) {
            cout << "Key not found in the list" << endl;
            return;
        }
        new_node->next = temp->next;
        temp->next = new_node;
    }

    void deleteNode(int key) {
        Node* temp = head;
        Node* prev = nullptr;

        if (temp != nullptr && temp->data == key) {
            head = temp->next;
            delete temp;
            return;
        }

        while (temp != nullptr && temp->data != key) {
            prev = temp;
            temp = temp->next;
        }

        if (temp == nullptr) {
            cout << "Key not found in the list" << endl;
            return;
        }

        prev->next = temp->next;
        delete temp;
    }

    void printList() {
        Node* temp = head;
        while (temp != nullptr) {
            cout << temp->data << "-";
            temp = temp->next;
        }
        cout << endl;
    }
};

int main() {
    LinkedList list;

    // Insert nodes
    list.insertAtEnd(1);
    list.insertAtEnd(2);
    list.insertAtEnd(3);
    list.insertAtEnd(4);

    cout << "Initial List: ";
    list.printList();

    cout << "After inserting in mid: ";
    list.insertInMid(5, 6);
    list.printList();

    // Delete a node
    cout << "After deleting 3: ";
    list.deleteNode(2);
    list.printList();

    // Insert at beginning
    cout << "After inserting 5 at beginning: ";
    list.insertAtBeginning(0);
    list.printList();

    return 0;
}